doc/tags_param.md

```diff
@@ -0,0 +1,205 @@
+gkernel.has<int32_t>("log_width");      // returns false
+gkernel.has<float32_t>("log_width");    // returns true
+```
+
+#### list_params()
+
+Coming soon.
+
+We discussed about how to use the new parameter framework in C++. Now let's look at one of the high level language that Shogun supports like Python.
+
+### Python API
+
+Let's look at another `GaussianKernel` example.
+
+```python
+# import shogun
+import modshogun as sg
+# Instantiate a Gaussian Kernel object
+gkernel = sg.GaussianKernel()
+# set width = 5
+gkernel.sets("log_width", 5.0)
+# if we want to use a Tag object
+float_width_tag = TagFloat("log_width")
+gkernel.sets(float_width_tag, 6.0)
+# get value of width
+w = gkernel.getsFloat("log_width")
+w = gkernel.getsInt("log_width")    # throws exception
+# or by using a Tag object
+w = gkernel.gets(float_width_tag)
+# to check if a parameter (corresponding to name) exists
+gkernel.has("log_width")    # returns true
+gkernel.has("foo")      # returns false
+# to check if a parameter (corresponding to name and type) exists
+gkernel.hasInt("log_width")     # returns false
+gkernel.hasFloat("log_width")   # returns true
+gkernel.hasFloat("foo")     # returns false
+```
+
+The semantics remain the same with minute syntactic differences. Important things to note:
+- `sets()` can be used to set a parameter using its name or a `Tag<Type>` (like `TagFloat`) object.
+- `gets()` can be used to query a parameter's value using a `Tag<Type>` (like `TagFloat`) object **but** `gets<Type>()` (like `getsFloat()`) is used to query the value by the parameter's name.
+- `has()` can be used to check the availability of a parameter by its name. `has<Type>()` (like `hasFloat()`) is used to check the availability of a parameter by it's name and type.
+
+These nomenclature of functions `sets`/`gets`/`has` is same in all the high level languages supported by Shogun i.e. Python, Ruby, R, Lua, Java, C# and Octave.
+
+*A pro tip:* Use `gets<Tab>`/`has<Tab>`/`Tag<Tab>` auto-completion to list all the types available.
+
+### For developers
+
+All this is accomplished by having a `std::unordered_map<BaseTag, Any>` (or `std::map<BaseTag, Any>` depending on the availability of C++11) in `SGObject.cpp`, where the map stores the values of parameters available in a Shogun class derived from `CSGObject`. From here on, I'll refer to the `SGObject`'s `std::unordered_map`/`std::map` as `map`.
+
+#### BaseTag
+
+`BaseTag` is the parent class of `Tag`. For a particular parameter, `BaseTag` stores the name, while `Tag` (being a template class) stores the type information. Name and type information are stored separately because `BaseTag` also stores a hash generated from the parameter's name. `BaseTag` is used as the key in the `map`, this results in fast look-ups.
+
+See examples of `BaseTag` in `SGObject.h`.
+
+#### Any
+
+`Any` allows to store objects of arbitrary types in a type agnostic way. This makes it possible to store a variety of types in the `map`.
```

Comment:
@OXPHOS I just realized that I need to talk about `NonOwningValueAnyPolicy` and `PointerValueAnyPolicy`, so will do that soon.